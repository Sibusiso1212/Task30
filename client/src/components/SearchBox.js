/**
 * Component for searching GitHub users.
 */
// Component for searching GitHub users
import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import '../styles/SearchBox.css';

const SearchBox = () => {
    // State management for search functionality
    const [query, setQuery] = useState('');
    const [results, setResults] = useState([]);
    const [loading, setLoading] = useState(false);
    const navigate = useNavigate();

    // Handle search form submission
    const handleSearch = async (e) => {
        e.preventDefault();
        if (!query.trim()) return;
        
        setLoading(true);
        try {
            // Make API call to search users
            const response = await fetch(`http://localhost:5000/api/search/users/${encodeURIComponent(query)}`);
            const data = await response.json();
            console.log('Search results:', data);
            setResults(data.items || []);
        } catch (error) {
            console.error('Search error:', error);
            setResults([]);
        }
        setLoading(false);
    };

    return (
        <div className="search-container">
            {/* Search form */}
            <form onSubmit={handleSearch}>
                <input
                    type="text"
                    value={query}
                    onChange={(e) => setQuery(e.target.value)}
                    placeholder="Search GitHub users..."
                    required
                    minLength={1}
                />
                <button type="submit" disabled={loading}>
                    {loading ? 'Searching...' : 'Search'}
                </button>
            </form>
            
            {/* Loading indicator */}
            {loading && <div className="loader">Loading...</div>}
            
            {/* Search results */}
            <div className="results">
                {results.length > 0 ? (
                    results.map(user => (
                        <div 
                            key={user.id} 
                            className="user-card"
                            onClick={() => navigate(`/user/${user.login}`)}
                        >
                            <img src={user.avatar_url} alt={user.login} />
                            <h3>{user.login}</h3>
                        </div>
                    ))
                ) : (
                    query && !loading && <p>No users found</p>
                )}
            </div>
        </div>
    );
};

export default SearchBox;
