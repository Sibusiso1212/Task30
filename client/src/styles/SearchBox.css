.search-container {
    max-width: 1200px;
    margin: 2rem auto;
    padding: 0 2rem;
}

.search-container form {
    display: flex;
    gap: 1rem;
    margin-bottom: 2rem;
}

.search-container input {
    flex: 1;
    padding: 1rem 1.5rem;
    font-size: 1.1rem;
    border: 2px solid #e2e8f0;
    border-radius: 12px;
    background-color: white;
    transition: all 0.3s ease;
}

.search-container input:focus {
    border-color: var(--accent-color);
    outline: none;
    box-shadow: 0 0 0 3px rgba(124, 58, 237, 0.1);
}

.search-container button {
    padding: 1rem 2rem;
    background: linear-gradient(135deg, #22c55e 0%, #16a34a 100%);
    color: white;
    border: none;
    border-radius: 12px;
    font-weight: 600;
    cursor: pointer;
    transition: transform 0.2s ease;
}

.search-container button:hover {
    transform: translateY(-2px);
}

.results {
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
    gap: 2rem;
}

.user-card {
    background: white;
    border-radius: 16px;
    padding: 1.5rem;
    text-align: center;
    box-shadow: var(--card-shadow);
    transition: all 0.3s ease;
}

.user-card:hover {
    transform: translateY(-5px);
    box-shadow: 0 20px 25px -5px rgb(0 0 0 / 0.1);
}

.user-card img {
    width: 120px;
    height: 120px;
    border-radius: 50%;
    border: 4px solid var(--secondary-color);
    margin-bottom: 1rem;
}

.user-card h3 {
    color: var(--text-primary);
    font-size: 1.25rem;
    margin: 0.5rem 0;
}

.loader {
      text-align: center;
      padding: 20px;
      font-size: 18px;
}
