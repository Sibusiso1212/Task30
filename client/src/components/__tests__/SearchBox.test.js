import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import SearchBox from '../SearchBox';

// Mock fetch function
global.fetch = jest.fn();

describe('SearchBox Component', () => {
    // Snapshot test
    test('matches snapshot', () => {
        const { container } = render(
            <BrowserRouter>
                <SearchBox />
            </BrowserRouter>
        );
        expect(container).toMatchSnapshot();
    });

    // Input field test
    test('updates input value on change', () => {
        render(
            <BrowserRouter>
                <SearchBox />
            </BrowserRouter>
        );
        const input = screen.getByPlaceholderText('Search GitHub users...');
        fireEvent.change(input, { target: { value: 'test' } });
        expect(input.value).toBe('test');
    });

    // Search functionality test
    test('performs search on form submission', async () => {
        const mockData = {
            items: [
                { id: 1, login: 'user1', avatar_url: 'url1' },
                { id: 2, login: 'user2', avatar_url: 'url2' }
            ]
        };

        global.fetch.mockResolvedValueOnce({
            json: () => Promise.resolve(mockData)
        });

        render(
            <BrowserRouter>
                <SearchBox />
            </BrowserRouter>
        );

        const input = screen.getByPlaceholderText('Search GitHub users...');
        const button = screen.getByText('Search');

        fireEvent.change(input, { target: { value: 'test' } });
        fireEvent.click(button);

        await waitFor(() => {
            expect(screen.getByText('user1')).toBeInTheDocument();
        });
    });
});