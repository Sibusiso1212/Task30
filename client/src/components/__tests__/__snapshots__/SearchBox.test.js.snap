// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox Component matches snapshot 1`] = `
<div>
  <div
    class="search-container"
  >
    <form>
      <input
        minlength="1"
        placeholder="Search GitHub users..."
        required=""
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Search
      </button>
    </form>
    <div
      class="results"
    />
  </div>
</div>
`;
